Adding Atom feed

Lucas-C opened the pull request:
This is an early version.
I need to add some rspec tests, + some feed validation using feedvalidator.org or validator.w3.org/feed/check.cgi

Note that this is the very first time I'm coding in Ruby ^^
I'm open to ANY feedback: code style, idioms, separation of concerns in Rails...

Lucas-C commented:
What it looks like:

![techlist_feed](https://cloud.githubusercontent.com/assets/925560/6314754/5beffb00-b9eb-11e4-93c7-17f059ee95af.png)

scharrier commented:
Thx ! I'll try to find some time to review it during the week.

scharrier commented:
Sorry for the delay. First, thanks a lot for your contribution !

I though about how we should add this atom feed, and I think the API would be a better place, even if it can be a surprising choice. Data presented in the feed is exactly the same that in the places endpoint (`api/v1/places.json`), except that it's not the same format, and places aren't ordered on the same attribute. With this in mind, it totally makes sense (for me, at least), to use `respond_to` in `controllers/api/v1/places#index` to handle the Atom response.

We would then be able to have `/api/v1/places.json` and `/api/v1/places.atom` endpoints, which could evolve like the rest of the API.

Thoughts ? ping @Lucas-C

scharrier commented on app/controllers/places_controller.rb:
```diff
@@ -2,6 +2,12 @@ class PlacesController < ApplicationController
   def index
   end
 
+  def feed
```
See https://github.com/mainelabs/techlist/pull/59#issuecomment-100263040. Could be moved to a `respond_to` in `api/v1/places#index`.

scharrier commented on spec/features/a_user_can_subscribe_to_the_place_feed.rb:
```diff
@@ -0,0 +1,44 @@
+
+feature 'A user subscribe to the place Atom feed' do
+  scenario 'and consult the complete correct feed page' do
+    CsvImportService.new('Place', Rails.root.join('spec/fixtures/csv/valid.csv'), state: 'active').save
```
We shouldn't load fixtures in specs, but use FactoryGirl to create objects needed for the tests.

See https://github.com/mainelabs/techlist/blob/master/spec/features/admin/an_admin_can_accept_a_new_place_spec.rb#L5, for instance.

scharrier commented on spec/features/a_user_can_subscribe_to_the_place_feed.rb:
```diff
@@ -0,0 +1,44 @@
+
+    expect(page).to have_title(t('feed.title'))
+
+    page.assert_selector('entry author name', count: 2)
```
Using factory girl, we could test directly the presence of our authors (but `author_name` is a private property we shouldn't give access to)

scharrier commented on spec/features/a_user_can_subscribe_to_the_place_feed.rb:
```diff
@@ -0,0 +1,44 @@
+    expect(page).to have_title(t('feed.title'))
+
+    page.assert_selector('entry author name', count: 2)
+    all('entry author name').each { |name| expect(name.text).to_not eq('') }
```
We should avoir loops and keep specs as simple as possible (one line = one test). If you remove fixtures and only generate 1 or 2 places, you should be able to test it more easily.

scharrier commented on spec/features/a_user_can_subscribe_to_the_place_feed.rb:
```diff
@@ -0,0 +1,44 @@
+    expect(all('updated').length).to eq(3)
+    all('updated').each { |updated| expect(updated.text).to_not eq('') }
+
+    expect(Place.first.updated_at).to_not eq(Place.last.updated_at)
```
Are you testing here that your CSV import worked here ?

scharrier commented on spec/features/a_user_can_subscribe_to_the_place_feed.rb:
```diff
@@ -0,0 +1,44 @@
+  end
+
+  scenario 'and the last updated place appears first' do
+    CsvImportService.new('Place', Rails.root.join('spec/fixtures/csv/valid.csv'), state: 'active').save
```
Same as https://github.com/mainelabs/techlist/pull/59/files#r29946817

scharrier commented on spec/features/a_user_can_subscribe_to_the_place_feed.rb:
```diff
@@ -0,0 +1,44 @@
+    expect(top_entry_title).to eq('Ecolutis')
+
+    # Updating 'Craftsmen Angers' place to make it come at the top
+    oldest_updated_place = Place.order('updated_at DESC').last
```
We shouldn't need this. You could have something simpler, like:

```ruby
last_place = create(:place, updated_at: 3.days.ago, name: 'Last entry')
first_place = create(:place, updated_at: DateTime.current, name: 'First entre')

expect(find(:xpath, '//feed/entry[1]/title').text).to eq('First entry')
expect(find(:xpath, '//feed/entry[2]/title').text).to eq('Last entry')
```

scharrier commented on spec/fixtures/csv/valid.csv:
```diff
@@ -1,2 +1,2 @@
 name;kind;street;zip_code;city;url;latitude;longitude;twitter_name;logo_url;description;owner_name;owner_email
-Craftsmen;company;25 rue Lenepveu;49000;Angers;http://craftsmen.io;47.472041;-0.551269;craftsmenhq;;;;
+Craftsmen Angers;company;25 rue Lenepveu;49100;Angers;http://craftsmen.io;47.472041;-0.551269;craftsmenhq;https://pbs.twimg.com/profile_images/425256684244566016/N0wcdLyQ_400x400.jpeg;Webdesign studio;Sébastien Charrier;sebastien@craftsmen.io
```
Shouldn't be needed if using factories instead of fixtures.

scharrier commented on spec/fixtures/csv/valid2.csv:
```diff
@@ -0,0 +1,2 @@
+name;kind;street;zip_code;city;url;latitude;longitude;twitter_name;logo_url;description;owner_name;owner_email
```
Same as https://github.com/mainelabs/techlist/pull/59/files#r29947819

scharrier commented on app/views/places/feed.atom.builder:
```diff
@@ -0,0 +1,37 @@
+def place_to_html_description(place)
```
We shouldn't define a method in a view. We could use a partial for this, or maybe a presenter.

Lucas-C commented:
My turn to be late to answer :) Sorry !

Ok, except from your comment on not using a function in a view, I think I addressed all the points you raised.
That is essentially moving the atom feed management into the controller api namespace and using FactoryGirl to improve the tests.

For [the remaining issue with the function definition in the view](https://github.com/mainelabs/techlist/pull/59#discussion_r29948095):
- why is that a problem ? It is only doing output formatting
- adding a presenter class or defining an equivalent partial function in the model looks like bloat code to me. Do you have any other suggestion ?
